nucliadb/nucliadb/purge/orphan_shards.py

```diff
@@ -0,0 +1,252 @@
+    driver: Driver,
+):
+    async with driver.transaction(read_only=True) as txn:
+        for shard_id, location in shards.items():
+            kb_exists = await KnowledgeBox.exist_kb(txn, location.kbid)
+            if kb_exists:
+                msg = "Found orphan shard for existing KB"
+            else:
+                msg = "Found orphan shard for already removed KB"
+
+            logger.debug(
+                msg,
+                extra={
+                    "shard_id": shard_id,
+                    "kbid": location.kbid,
+                    "node_id": location.node_id,
+                },
+            )
+
+
+async def purge_orphan_shards(driver: Driver):
+    orphan_shards = await detect_orphan_shards(driver)
+
+    unavailable_nodes: dict[str, datetime] = {}
+    for shard_id, location in orphan_shards.items():
+        node = manager.get_index_node(location.node_id)
+        if node is None:
+            now = datetime.now()
+            last_try = unavailable_nodes.setdefault(location.node_id, now)
+            if (now - last_try).seconds < 10:
+                logger.warning(
+                    "Node not available while purging orphan shards, skipping",
+                    extra={"node_id": location.node_id},
+                )
+                unavailable_nodes[location.node_id] = now
+            continue
+
+        logger.info(
+            "Deleting orphan shard from index node",
+            extra={
+                "shard_id": shard_id,
+                "kbid": location.kbid,
+                "node_id": location.node_id,
+            },
+        )
+        await node.delete_shard(shard_id)
+
+
+def parse_arguments():
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "--purge",
+        action="store_true",
+        default=False,
+        required=False,
+        help="Purge detected orphan shards",
+    )
+    args = parser.parse_args()
+    return args
+
+
+async def main():
+    """This script will detect orphan shards, i.e., indexed shards with no
+    reference in our source of truth.
+
+    Purging a knowledgebox can lead to orphan shards if index nodes are not
+    available or fail. It's possible that other procedures in our database left
+    orphan shards too.
+
+    ATENTION! In the future, some new process which adds new shards could be
+    implemented. If orphan shard detection is not updated, this will lead to
+    incorrect detection. To avoid data loss/corruption, it is highly recommended
+    to don't remove orphan shards that hasn't exist for a long time.
+
+    """
+    args = parse_arguments()
+
+    await setup_cluster()
+    driver = await setup_driver()
+
+    try:
+        orphan_shards = await detect_orphan_shards(driver)
+        logger.info(f"Orphan shards detect found {len(orphan_shards)} orphans")
+        await report_orphan_shards(orphan_shards, driver)
+        if args.purge:
+            await purge_orphan_shards(driver)
```

Comment:
If `purge=True` we will be running `detect_orphan_shards`.
you could change it to something like:

```
if args.purge:
   await purge_orphan_shards()
else:
   await report_orphan_shards()
```
And each internall calls detect_orphan_shards